lib/vanagon/component.rb

```diff
@@ -190,16 +192,74 @@ def configfiles
       @files.select(&:configfile?)
     end
 
+    # @return [Set] a list of unique mirror URIs that should be used to
+    #   retrieve the upstream source before attempting to retrieve from
+    #   whatever URI was defined for #uri.
+    def mirrors
+      @mirrors ||= Set.new [deprecated_rewrite_url].compact
+    end
+
+    # Retrieve upstream source file from a mirror, by randomly iterating
+    # through #mirrors until there's no more mirrors left. Will #warn
+    # if the mirror's URI cannot be resolved or if the URI cannot
+    # be retrieved. Does not suppress any errors from
+    # Vanagon::Component::Source.
+    #
+    # @return [Boolean] return True if the source can be retrieved,
+    #   or False otherwise. This is because because each subclass of
+    #   Vanagon::Component::Source returns an inconsistent value
+    #   if #fetch is successful.
+    def fetch_mirrors(options)
+      mirrors.to_a.shuffle.each do |mirror|
```

Comment:
Mirrors have no precedence, and hitting one should be equivilent to hitting any of them. So using `Array#shuffle` ensures that the order is randomized, but that the entire array is still worked through.

![26412689](https://cloud.githubusercontent.com/assets/344926/24875864/6599468a-1dde-11e7-9f4f-58656ed29a70.jpg)